pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpartition/SegmentPartitionMetadataManager.java

```diff
@@ -0,0 +1,189 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.broker.routing.segmentpartition;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.zookeeper.datamodel.ZNRecord;
+import org.apache.pinot.broker.routing.segmentmetadata.SegmentZkMetadataFetchListener;
+import org.apache.pinot.spi.utils.CommonConstants;
+
+
+/**
+ * The {@code PartitionDataManager} manages partitions of a table. It manages
+ *   1. all the online segments associated with the partition and their allocated servers
+ *   2. all the replica of a specific segment.
+ * It provides API to query
+ *   1. For each partition ID, what are the servers that contains ALL segments belong to this partition ID.
+ *   2. For each server, what are all the partition IDs and list of segments of those partition IDs on this server.
+ */
+public class SegmentPartitionMetadataManager implements SegmentZkMetadataFetchListener {
+  private final String _tableNameWithType;
+
+  // static content, if anything changes for the following. a rebuild of routing table is needed.
+  private final String _partitionColumn;
+  private final String _partitionFunctionName;
+  private final int _numPartitions;
+
+  // cache-able content, only follow changes if onlineSegments list (of ideal-state) is changed.
+  private final Map<String, List<String>> _segmentToOnlineServersMap = new HashMap<>();
+  private final Map<String, Integer> _segmentToPartitionMap = new HashMap<>();
+
+  // computed value based on status change.
+  private Map<Integer, Set<String>> _partitionToFullyReplicatedServersMap;
+  private Map<Integer, List<String>> _partitionToSegmentsMap;
+
+  public SegmentPartitionMetadataManager(String tableNameWithType, String partitionColumn, String partitionFunctionName,
+      int numPartitions) {
+    _tableNameWithType = tableNameWithType;
+    _partitionColumn = partitionColumn;
+    _partitionFunctionName = partitionFunctionName;
+    _numPartitions = numPartitions;
+  }
+
+  @Override
+  public void init(IdealState idealState, ExternalView externalView, List<String> onlineSegments,
```

Comment:
review note: when `init` and `onAssignmentChange` it is actually 2-step
1. update with zk metadata (for partition mapping) -- done only on those that have zk metadata changed
2. update with external vew (for online server mapping) -- done on every online segment